Prevent headers which are signed by slashed validators

hewison-chris opened the pull request:
Fixes #3296. 
First commit ensures that we would not accept a header in this situation (as we receive it from other nodes).
The second commit  prevents the situation where these headers are generated.

hewison-chris committed:
Header validation: Accept only consistent headers

If a signature is included for a validator then the preimage for that
validator must also be included.

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1096,6 +1096,12 @@ extern(D):
+                {
+                    log.trace("{}: Adding missing preimage for {} at height {}",
+                        __FUNCTION__, val.address, header.height);
+                    header.preimages[idx] = val.preimage[header.height];
```
Is there any possibility where `header.height` is greater than the` val.preimage.height`? I wonder if we should check that.

hewison-chris replied:
This is removed now as it does not make sense to add pre-images at this point.

linked0 reviewed (commented):
Broadly LGTM but one question.

omerfirmak reviewed (changes requested):
We should never add a preimage to a header. They indicate that we slashed a validator during consensus and that fact should never change after externalization of the slot.

Also they are part of the block/blockheader hash, so doing so would invalidate any previous signature.

What we  probably should just do is not accept the signature from a slashed validator, even if we get their preimage after the block externalization.

hewison-chris commented:
> We should never add a preimage to a header. They indicate that we slashed a validator during consensus and that fact should never change after externalization of the slot.
> 
> Also they are part of the block/blockheader hash, so doing so would invalidate any previous signature.
> 
> What we probably should just do is not accept the signature from a slashed validator, even if we get their preimage after the block externalization.

Yes I see your point, the block should be considered immutable at this point except for the signature and validators bitmask header fields, which are not included in the hash. 🤦

hewison-chris committed:
Only add signature if not slashed

If a block is being externalized that does not contain a preimage for a
validator we must not add a signature for that validator. The validator
is being slashed at this point so it is too late to add the preimage and
signature, otherwise the block hash would change and the combined
signature would become invalid.

linked0 reviewed (approved):
LGTM

linked0 commented:
I'm running the validator built with these commits and continuously getting the error message as follows. The slashed validator is my validator anyway.
```
2022-04-15 06:24:14,87 Info [agora.network.Manager] - Received blocks [282..305]
2022-04-15 06:24:14,93 Error [agora.consensus.state.Ledger] - Block#282: Header is not consistent. The signature at index 23 is included for slashed validator boa1xz7hrkcjgecknn25npj5y7pkqtcw0qn48mvzddrur9clnq7zdenfjmjnpj8
```
I think we should re-run the Testnet.
